File: python/src/falcon_core/autotuner_interfaces/interpretations/interpretation_context.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.interpretation_context import InterpretationContext as _CInterpretationContext
from falcon_core.math.axes import Axes
from falcon_core.generic.list import List
from falcon_core.autotuner_interfaces.contexts.measurement_context import MeasurementContext
from falcon_core.physics.units.symbol_unit import SymbolUnit

class InterpretationContext:
    """Python wrapper for InterpretationContext."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def new(cls, independant_variables: Axes, dependant_variables: List, unit: SymbolUnit) -> InterpretationContext:
        return cls(_CInterpretationContext.new(independant_variables._c, dependant_variables._c, unit._c))

    @classmethod
    def from_json(cls, json: str) -> InterpretationContext:
        return cls(_CInterpretationContext.from_json(json))

    def independent_variables(self, ) -> Axes:
        ret = self._c.independent_variables()
        if ret is None: return None
        return Axes(ret)

    def dependent_variables(self, ) -> List:
        ret = self._c.dependent_variables()
        if ret is None: return None
        return List(ret)

    def unit(self, ) -> SymbolUnit:
        ret = self._c.unit()
        if ret is None: return None
        return SymbolUnit._from_capi(ret)

    def dimension(self, ) -> None:
        ret = self._c.dimension()
        return ret

    def dependent_variable(self, variable: MeasurementContext) -> None:
        ret = self._c.dependent_variable(variable._c)
        return ret

    def replace_dependent_variable(self, index: Any, variable: MeasurementContext) -> None:
        ret = self._c.replace_dependent_variable(index, variable._c)
        return ret

    def get_independent_variables(self, index: Any) -> MeasurementContext:
        ret = self._c.get_independent_variables(index)
        if ret is None: return None
        return MeasurementContext._from_capi(ret)

    def with_unit(self, unit: SymbolUnit) -> InterpretationContext:
        ret = self._c.with_unit(unit._c)
        return cls._from_capi(ret)

    def equal(self, b: InterpretationContext) -> None:
        ret = self._c.equal(b._c)
        return ret

    def not_equal(self, b: InterpretationContext) -> None:
        ret = self._c.not_equal(b._c)
        return ret

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, InterpretationContext):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, InterpretationContext):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/_opam/.opam-switch/sources/falcon_core/src/falcon_core/autotuner_interfaces/interpretations/interpretationcontext.ml
open Ctypes
open Capi_bindings
open ErrorHandling

open Falcon_core.Autotuner_interfaces.Contexts
open Falcon_core.Generic
open Falcon_core.Math
open Falcon_core.Physics.Units

class c_interpretationcontext (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.interpretationcontext_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module InterpretationContext = struct
  type t = c_interpretationcontext

  let copy (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let ptr = Capi_bindings.interpretationcontext_copy handle#raw in
      ErrorHandling.raise_if_error ();
      new c_interpretationcontext ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.interpretationcontext_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_interpretationcontext ptr

  let make (independant_variables : AxesMeasurementContext.t) (dependant_variables : ListMeasurementContext.t) (unit : SymbolUnit.t) : t =
    ErrorHandling.multi_read [independant_variables; dependant_variables; unit] (fun () ->
      let ptr = Capi_bindings.interpretationcontext_create independant_variables#raw dependant_variables#raw unit#raw in
      ErrorHandling.raise_if_error ();
      new c_interpretationcontext ptr
    )

  let equal (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.interpretationcontext_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.interpretationcontext_not_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.interpretationcontext_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let independentVariables (handle : t) : AxesMeasurementContext.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.interpretationcontext_independent_variables handle#raw in
      ErrorHandling.raise_if_error ();
      new c_axesmeasurementcontext result
    )

  let dependentVariables (handle : t) : ListMeasurementContext.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.interpretationcontext_dependent_variables handle#raw in
      ErrorHandling.raise_if_error ();
      new c_listmeasurementcontext result
    )

  let unit (handle : t) : SymbolUnit.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.interpretationcontext_unit handle#raw in
      ErrorHandling.raise_if_error ();
      new c_symbolunit result
    )

  let dimension (handle : t) : int =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.interpretationcontext_dimension handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let addDependentVariable (handle : t) (variable : MeasurementContext.t) : unit =
    ErrorHandling.multi_read [handle; variable] (fun () ->
      let result = Capi_bindings.interpretationcontext_add_dependent_variable handle#raw variable#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let replaceDependentVariable (handle : t) (index : int) (variable : MeasurementContext.t) : unit =
    ErrorHandling.multi_read [handle; variable] (fun () ->
      let result = Capi_bindings.interpretationcontext_replace_dependent_variable handle#raw index variable#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let getIndependentVariables (handle : t) (index : int) : MeasurementContext.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.interpretationcontext_get_independent_variables handle#raw index in
      ErrorHandling.raise_if_error ();
      new c_measurementcontext result
    )

  let withUnit (handle : t) (unit : SymbolUnit.t) : t =
    ErrorHandling.multi_read [handle; unit] (fun () ->
      let result = Capi_bindings.interpretationcontext_with_unit handle#raw unit#raw in
      ErrorHandling.raise_if_error ();
      new c_interpretationcontext result
    )

end